tokio-util/src/udp/frame.rs

```diff
@@ -57,7 +57,7 @@ where
     T: Borrow<UdpSocket>,
     C: Decoder,
 {
-    type Item = Result<(C::Item, SocketAddr), C::Error>;
+    type Item = Result<(C::Item, SocketAddr), (C::Error, Option<SocketAddr>)>;
```

Comment:
This will make it so`?` can't be used right?

Reply:
Yes, it also breaks the tests currently which assume that TryFuture::Error is the same for both the send and recv operations.  I knew this would be the case but wanted to highlight the issues from the task in this PR and collaborate a little bit on how to solve things elegantly.  Again the most straight forward approach is to just fix the tests to not make this assumption and expect that everyone's code breaks.  I think I currently prefer this solution as it does the best job of explaining to developers what was wrong with the old API and asks them to be thoughtful about restructuring their own code (that is, developers were probably silently swallowing decode errors from peers).

An alternative might be a newtype which impls From<C::Error> so callers can just ignore the Option<SocketAddr> we added if they want.  Unfortunately I think this has the greatest chance of developers just papering over the problem and ignoring the new information.  This would also still require some level of breaking change as the try_join in the tests still doesn't work because the types aren't the same (even if there's a From relationship that could make them the same).

Input welcome, I feel a little bit out of my depth on such a hugely popular library as I'm a bit new to Rust myself but definitely willing to learn from the experts :)